bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java

```diff
@@ -0,0 +1,409 @@
+    }
+
+    private State getTotalConsumption() {
+        return getQuantity("GetPMWarningThreshold", "TotalConsumption", "2", SmartHomeUnits.KILOWATT_HOUR);
+    }
+
+    private State getTemperature() {
+        return getQuantity("GetCurrentTemperature", "CurrentTemperature", "3", SIUnits.CELSIUS);
+    }
+
+    private State getState() {
+        try {
+            String state = soapAction("GetSocketSettings", "OPStatus", moduleParameters("1")).toLowerCase();
+            if ("true".contentEquals(state)) {
+                return OnOffType.ON;
+            } else if ("false".equals(state)) {
+                return OnOffType.OFF;
+            } else {
+                logger.error("Unexpected state: {}", state);
+            }
+        } catch (CommunicationException e) {
+            logger.error("Failed to get state (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    private void setState(final OnOffType state) {
+        try {
+            final String encodedState = OnOffType.ON.equals(state) ? "true" : "false";
+            final String controlParameters = moduleParameters("1")
+                    + "<NickName>Socket 1</NickName><Description>Socket 1</Description>" + "<OPStatus>" + encodedState
+                    + "</OPStatus>";
+            final String response = soapAction("SetSocketSettings", "SetSocketSettingsResult", controlParameters);
+            if (!"ok".equals(response.toLowerCase())) {
+                logger.error("Failed to set state.");
+            }
+        } catch (Exception e) {
+            // pass
+        }
+    }
+
+    private String moduleParameters(final String module) {
+        return "<ModuleID>" + module + "</ModuleID>";
+    }
+
+    private State getQuantity(final String action, final String name, final String module, final Unit<?> unit) {
+        try {
+            final String s = soapAction(action, name, moduleParameters(module));
+            try {
+                if (s == null) {
+                    return UnDefType.UNDEF;
+                } else {
+                    return new QuantityType<>(Double.valueOf(s), unit);
+                }
+            } catch (NumberFormatException e) {
+                logger.error("Unexpected number format: {}", s);
+            }
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    private String soapAction(final String action, final String name, final String params)
+            throws CommunicationException {
+        Document document = post(action, requestBody(action, params), authentication);
+        return getElementByTagName(document, name);
```

Comment:
I still think you should attempt to use the existing communication class to avoid duplicate code. Changing this method to something like the following would allow you to remove a lot of the duplication (apologies for the lack of code formatting which I can't get to work on my phone):

`actionMsg = messageFactory.createMessage();
actionMsg.getSOAPHeader().detachNode();
final SOAPBody soapBody = actionMsg.getSOAPBody();
final SOAPElement soapBodyElem = soapBody.addChildElement(action, "", HNAP_XMLNS);

// Add module parameters here, but it might be easier just to add this code to setState and getQuantity methods and remove this method.

final MimeHeaders headers = actionMsg.getMimeHeaders();
final String actionUrl = "\"http://purenetworks.com/HNAP1/" + action + "\"";
headers.addHeader(SOAPACTION, actionUrl);

setAuthenticationHeaders(actionMsg);

final Document document = sendReceive(actionMsg, 5000);

return getElementByTagName(document, name);`

Reply:
I agree that I should use the existing SOAP implementation. I will change that once the long-term test has finished.